### DISCUSSION TITLE
Performance of iterative solvers with mesh refinement / hanging nodes

### DISCUSSION THREAD

**Discussion by matthiasneuner** [idaholab/moose#20344] [Q&A General] [answered]
Hi, we recently tried to use an initial local mesh refinement (h refinement) for our solid mechanics based applications, using MOOSE's refinement capabilities. However, what we observe is a considerably degrading performance of the GMRES/BoomerAMG combo, i.e., the linear solver takes many iterations or does not converge at all within the defined limit of linear iterations.
I assume that this is related to some constraints (of which nature are they?) introduced by the hanging nodes of the non-compatible finite element meshes. 

Is this the usual behavior, and are there any approaches to tackle those performance issues?

Thank you and best regards!

**Comment by GiudGiud** [comment]
Hello

I ll let @roystgnr comment about the hanging nodes, I m not aware of that.
A few things:
- you can let the the linear solver converge by increasing the `l_max_its` parameter.
- what are your executioner parameters? Maybe we can improve the preconditioning / solver combo
- do you really need to refine the whole domain? If not then use adaptivity in certain subdomains only or use the RefineBlockGenerator.
- do you need to refine from the beginning of the simulation? If not then start with a coarser mesh, converge that solution, then refine. 

Guillaume

**Comment by lindsayad** [comment, marked as answer]
Do you know what the magnitude of your physics Jacobians is? Is it of order Young's modulus which is often of order 1e9-1e12? Hanging node constraints do indeed introduce Jacobian rows of order unity, e.g.:
```c++
void DofMap::enforce_constraints_on_jacobian (const NonlinearImplicitSystem & system,
                                              SparseMatrix<Number> * jac) const
{
  parallel_object_only();

  if (!this->n_constrained_dofs())
    return;

  if (!jac)
    jac = system.matrix;

  libmesh_assert_equal_to (this, &(system.get_dof_map()));

  for (const auto & pr : _dof_constraints)
    {
      dof_id_type constrained_dof = pr.first;
      if (!this->local_index(constrained_dof))
        continue;

      const DofConstraintRow constraint_row = pr.second;

      for (const auto & j : constraint_row)
        jac->set(constrained_dof, j.first, -j.second);
      jac->set(constrained_dof, constrained_dof, 1);
    }
}
```
So this does have the potential to introduce scaling issues if your physics are dramatically different from order unity

**Reply by roystgnr** [reply]
This is a good hypothesis.  I've never noticed solver issues from hanging nodes, but most of my own work has been with nondimensionalized physics.  I'm kind of surprised that that's enough to throw BoomerAMG for a loop, though; generally it doesn't take too much of a preconditioner to notice that some rows/columns tend to be much bigger than others and to try to compensate accordingly.

**Comment by matthiasneuner** [comment]
Thank you to all your responses!

I think it might be related to the magnitude of the Jacobian; I am using a coupled Cosserat approach; hence I am also solving for the balance of angular momentum which might have a completely different magnitude. Furthermore, I am dealing with strongly localized solutions due to large deformations, which BoomerAMG doesn't really like anyways:
https://matthiasneuner.github.io/chamois/index.html

It is good to know that in principle hanging nodes should not cause such problems, this helps definitely a lot.
